# <mark>위클리퀘스트 3번</mark>

In [30]:
import torch
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from torch.utils.data import TensorDataset, DataLoader

In [3]:
<mark>이 글씨의 배경은 노란색이 됩니다.</mark>from sklearn.datasets import load_breast_cancer
import pandas as pd

# 데이터 불러오기
breast = load_breast_cancer()

# 데이터프레임으로 변환 (가독성 향상)
df = pd.DataFrame(breast.data, columns=breast.feature_names)

print("데이터셋 크기:", df.shape)

display(df.head())

데이터셋 크기: (569, 30)


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


In [10]:
#정답 데이터 가져오기

answer = breast.target
print(answer)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

In [12]:
# x,y 분리
# 학습/테스트 분리 : train_test_split

X = df
y = answer

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [28]:
# 표준화
# StandardScaler
# 표준화 : 각 열의 데이터를 모두 평균0, 표준편차 1로 변환
# --> 우리가 사용한 데이터가 어떤건 단위가 면적, 반지름, 매끄러움 등이라 값의 특성이 너무 다르다

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [17]:
from sklearn.linear_model import Perceptron
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB

In [18]:
# 4개의 모델 학습 및 예측

perceptron_model = Perceptron()
svc_model = SVC(kernel = 'linear')
random_forest_model = RandomForestClassifier()
naive_bayes_model = GaussianNB()

In [20]:
# 4개의 모델을 모두 동일한 데이터(지표)로 학습 진행

perceptron_model.fit(X_train, y_train )
svc_model.fit(X_train , y_train )
random_forest_model.fit(X_train , y_train )
naive_bayes_model.fit(X_train , y_train)

GaussianNB()

In [22]:
# 4개의 모델을 모두 동일한 데이터(지표)로 평가

answer1 = perceptron_model.predict(X_test)
answer2 = svc_model.predict(X_test)
answer3 = random_forest_model.predict(X_test)
answer4 = naive_bayes_model.predict(X_test)

In [25]:
from sklearn.metrics import accuracy_score

In [27]:
# 정확도에 대해 출력

accuracy1 = accuracy_score(y_test, answer1)
print('perceptron: ', accuracy1)

accuracy2 = accuracy_score(y_test ,answer2)
print('svc: ', accuracy2)

accuracy3 = accuracy_score(y_test, answer3)
print('random forest: ', accuracy3)

accuracy4 = accuracy_score(y_test ,answer4)
print('naive bayes: ', accuracy4)

perceptron:  0.9736842105263158
svc:  0.956140350877193
random forest:  0.956140350877193
naive bayes:  0.9649122807017544


 *   정답 y 가져오기— breast.target을 y로 지정합니다. (지금 df에는 입력만 있고 정답이 빠져 있음)
 *  X와 y 분리 확정 — X는 breast.data(또는 만든 df), y는 위에서 정한 target. 인코딩·더미·dropna 단계는 건너뜀
 *  학습/테스트 분리 — train_test_split으로 한 번만 나누기 (test 비율과 random_state 정해두기)
 *  표준화(스케일링) — StandardScaler 적용. train에 fit하고 그 기준으로 test를 변환 (Perceptron·SVM 때문에 필요)
 *  Perceptron 학습 및 예측
 *  SVM 학습 및 예측
 *  Random Forest 학습 및 예측
 *  Naive Bayes 학습 및 예측
 *  네 모델 동일 지표로 평가 — 같은 test 데이터로 accuracy 등 같은 기준 적용 (필요시 정밀도·재현율도)
 *  결과 비교 및 정리


 ## <font color="red">claude 와 대화</font>
 https://claude.ai/share/cd870ffe-11bf-404f-82c9-e9158f162bef

# <mark>위클리퀘스트 4번</mark>

In [31]:
from sklearn.datasets import make_classification

In [64]:
# 1. 가상 데이터 생성

X, y = make_classification(n_samples = 1000, n_classes = 2, weights = [0.9], n_features = 10, random_state = 42)
print('이진으로 결과 출력: ', np.unique(y))
print('데이터 총 길이: ', len(y))
print('클래스별 개수: ', np.unique(y, return_counts=True)) #각 정답(0,1)의 개수를 9:1 비율로 생성

이진으로 결과 출력:  [0 1]
데이터 총 길이:  1000
클래스별 개수:  (array([0, 1]), array([898, 102]))


In [65]:
# 2. train/test 분리
# 0.2로 분리하면서 각각
#0 - 898개의 0.8배 = 718개 / 1 - 102개의 0.8배 = 82개를 train 데이터로 가지게 됨

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42, stratify = y)

In [66]:
from sklearn.metrics import classification_report

In [67]:
# 3. 기준선 측정 (증강 x)

random_forest_model.fit(X_train, y_train)
before1 = random_forest_model.predict(X_test)
print(classification_report(y_test, before1))

              precision    recall  f1-score   support

           0       0.96      0.99      0.97       180
           1       0.86      0.60      0.71        20

    accuracy                           0.95       200
   macro avg       0.91      0.79      0.84       200
weighted avg       0.95      0.95      0.95       200



In [48]:
from imblearn.over_sampling import SMOTE

In [68]:
# 4. 증강 기법 적용
#증강 전에는 : 0 - 898개의 0.8배 = 718개 / 1 - 102개의 0.8배 = 82개를 train 데이터로 가지게 됨
#증강 후에는 : 답을 1로 가지는 데이터의 수도 0과 동일하게 718개로 맞춤

smote = SMOTE()
X_train_aug, y_train_aug = smote.fit_resample(X_train, y_train)

print(np.unique(y_train_aug, return_counts=True))

(array([0, 1]), array([718, 718]))


In [69]:
# 5. 실험군 측정 (증강 o)

random_forest_model.fit(X_train_aug, y_train_aug)
after1 = random_forest_model.predict(X_test)
print(classification_report(y_test, after1))

              precision    recall  f1-score   support

           0       0.98      0.98      0.98       180
           1       0.81      0.85      0.83        20

    accuracy                           0.96       200
   macro avg       0.90      0.91      0.90       200
weighted avg       0.97      0.96      0.97       200



In [74]:
# 6. 비교 및 해석
print('====== 증강 전 모델 성능 ======')
print(classification_report(y_test, before1), '\n')
print(np.unique(y, return_counts=True))
print('정확도(accuracy): ', accuracy_score(y_test, before1))


print('\n\n====== 증강 후 모델 성능 ======')
print(classification_report(y_test, after1))
print(np.unique(y_train_aug, return_counts=True))
print('정확도(accuracy): ', accuracy_score(y_test, after1))

====== 증강 전 모델 성능 ======
              precision    recall  f1-score   support

           0       0.96      0.99      0.97       180
           1       0.86      0.60      0.71        20

    accuracy                           0.95       200
   macro avg       0.91      0.79      0.84       200
weighted avg       0.95      0.95      0.95       200
 

(array([0, 1]), array([898, 102]))
정확도(accuracy):  0.95


====== 증강 후 모델 성능 ======
              precision    recall  f1-score   support

           0       0.98      0.98      0.98       180
           1       0.81      0.85      0.83        20

    accuracy                           0.96       200
   macro avg       0.90      0.91      0.90       200
weighted avg       0.97      0.96      0.97       200

(array([0, 1]), array([718, 718]))
정확도(accuracy):  0.965


<font color="blue">우리는 정답을 1로 하는 데이터를 적게 했고, 증강으로 늘렸기 때문에 거기만 본다.</font>


증강 전에는:
```
*   실제 1의 개수(support) : 20개
*   모델이 1이라고 찍은 것 중에 실제로 1인 비율(precision) : 0.86
*   실제 1인 20개 중 모델이 몇%를 잡았는지(recall) : 0.6
*   recall-precision 의 균형 점수(f1-score) : 0.71
```

증강 후에는:
```
*   실제 1의 개수(support) : 20개
*   모델이 1이라고 찍은 것 중에 실제로 1인 비율(precision) : 0.81
*   실제 1인 20개 중 모델이 몇%를 잡았는지(recall) : 0.85
*   recall-precision 의 균형 점수(f1-score) : 0.83
```


즉,

| 구분 | before | after |
| :--- | :---: | :---: |
| **정확도 (Accuracy)** | 0.95 | 0.965 |
| **정밀도 (precision)** | 0.86 | 0.81 |
| **재현율 (recall)** | 0.60 | 0.83 |

<br>

1.   1의 분석 데이터가 102개 -> 718개로 늘어나면서 헛다리(precision)이 조금 증가.
2.   하지만 데이터가 늘어나면서 1을 잡는 확룔이 (12/20)개에서 (17/20)개로 늘어났다.



 ## <font color="red">claude 와 대화 - 앞과 이이서 진행함</font>

https://claude.ai/share/cd870ffe-11bf-404f-82c9-e9158f162bef